sql/analyzer/resolve_columns.go

```diff
@@ -183,34 +236,524 @@ func dedupStrings(in []string) []string {
+				sq = e
+				return true
+			}
+			return false
+		})
+		if found {
+			return sq
+		}
+	}
+	return nil
+}
+
+// walk is a top-down traversal of the plan tree that builds a linked-list
+// of scope node namespace dependencies. Dependencies flow
+// downwards linearly from the tree root to plan.ResolvedTable leaves, and
+// in the future other sql.RelationalNode types. This is broad enough to
+// absorb several roles:
+//   1) This can qualify expressions if upon finding a namespace row source,
+//      we traverse upwards in the linked list bookkeeping columns between
+//      three buckets: i) unqualified, ii) qualified, iii) alias. Conflicts
+//      between multiple column sources would provide ambiguous column errors.
+//   2) Type checking -- when all of a node's columns are qualified, we can
+//      verify that node's expression type semantics.
+//   3) Filter pushdown - filters can move to their referenced relational
+//      nodes. In the case of multi-relation filters, they would move to the
+//      top-level join node that includes all referenced relationas.
+//   4) Column functional dependencies?
+//   3) Builder methods can be associated with specific optimization rules
+//      that only apply to specific nodes.
+func (b *builder) walk(n sql.Node, inScope *scope) *scope {
+	var outScope = inScope
+	transform.Inspect(n, func(n sql.Node) bool {
+		inScope = outScope
+
+		if sqe := findSubqueryExpr(n); sqe != nil {
+			inScope = b.buildSubqueryExpr(sqe, inScope)
+		}
+
+		switch n := n.(type) {
+		case *plan.ResolvedTable, *plan.IndexedTableAccess:
+			outScope = b.buildTable(n, inScope)
+		case *plan.Values:
+			outScope = b.buildValues(n, inScope)
+		case *plan.TableAlias:
+			outScope = b.buildTableAlias(n, inScope)
+		case *plan.Union:
+			outScope = b.buildUnion(n, inScope)
+			return false
+		case *plan.SubqueryAlias:
+			outScope = b.buildSubqueryAlias(n, inScope)
+		case plan.JoinNode, *plan.CrossJoin:
+			outScope = b.buildJoin(n, inScope)
+		case *plan.NaturalJoin:
+			outScope = b.buildNaturalJoin(n, inScope)
+		case *plan.GroupBy:
+			outScope = b.buildGenericNode(n, inScope)
+		case *plan.Filter, *plan.Project, *plan.Sort, *plan.Window:
+			outScope = b.buildGenericNode(n, inScope)
+		}
+		return true
+	})
+	return outScope
+}
+
 // getNodeAvailableSymbols returns the set of table and column names accessible to the node given and using the scope
 // given. Table aliases overwrite table names: the original name is not considered accessible once aliased.
 // The value of the map is the same as the key, just used for existence checks.
-func getNodeAvailableNames(n sql.Node, scope *Scope) availableNames {
-	names := make(availableNames)
-
+func getNodeAvailableNames(n sql.Node, scope *Scope, names availableNames, nestingLevel int) availableNames {
 	// Examine all columns, from the innermost scope (this one) outward.
-	getColumnsInNodes(n.Children(), names, 0)
+	getColumnsInNodes(n.Children(), names, nestingLevel)
```

Comment:
This is kind of busted, you've overloaded nesting level to where it doesn't mean any one thing here.

We start at some nesting level passed in, and then walk the scope setting nesting level = 1..N, which is going to overwrite the work outside the loop.

With these rewrites, this logic is really only appropriate when scope == nil, or when scope != nil and nestingLevel is 0. It should be two different methods to reflect these two different uses.

Reply:
I tipped the balance between qualify, resolve, and the group by rules. I tried duck taping instead of rewriting, which maybe wasn't the best idea. I am actively planning how to fix these three once we ship V1 and performance is less of a focus.